### ISSUE TITLE
Private registry / Configure npm registry for module download

### ISSUE THREAD

**Issue by geglock** [denoland/deno#16105]
I would like to be able to use a custom, self-hosted npm registry.
Especially for resolving company-local private packages based on _scopes_ this would be very helpful.
As far as I can see, currently only the npm public registry (https://registry.npmjs.org/) can be used.

Deno should support something similar to npm's configuration:
```
npm config set registry ...
npm config set @mycompany:registry ...
```

**Comment by jiawei397** [comment]
I think it is necessary. I want to pay attention to this progress, because in China, I sometimes download the resources of npm very slowly.

**Comment by jimisaacs** [comment]
+1 this is a basic requirement for enterprise use

**Comment by sithumonline** [comment]
Good to have GitLab private package registry support 🚀

**Comment by NickTomlin** [comment]
It looks like this is currently possible through the `DENO_NPM_REGISTRY ENV flag 

```
# execute a remote script ala `npx`
DENO_NPM_REGISTRY=https://my-npm-instance.com deno run npm:@my-internal-scope/my-internal-package
# Start module
cat index.ts

import * as myModule from "npm:@my-internal-scope/my-internal-package"
console.log(myModule)

DENO_NPM_REGISTRY=https://my-npm-instance.com deno run index.ts
```

**Comment by dsherret** [comment]
Sorry, the `DENO_NPM_REGISTRY` env variable wasn't intended to be stable at this point and was undocumented, but yeah it should work for now. I added a warning on its use in https://github.com/denoland/deno/pull/16953 in future versions, but just beware that it may be changed when we come up with a real solution for this problem (though likely it will still remain... it's just not something that's been discussed yet).

**Comment by NickTomlin** [comment]
@dsherret thanks for clarifying that and a adding that warning 😄 

Is this issue the best place for community feedback to try and help shape that work?

As @jimisaacs mentioned, I think this is crucial in bringing the magic of `npm:*` to the enterprise space (full disclosure: Jim and I work together at Netflix). Internally, the company has done a lot of investment in scaling, hardening, and auditing our private NPM registry. Being able to leverage that directly, instead of having to go through all the same technical and procedural hoops to stand up infrastructure to sit in front of it (e.g. `esm.sh`), would be huge win.

**Comment by dsherret** [comment]
We're going to discuss this issue tomorrow in our weekly meeting. For now I'm not adding the warning because the environment variable is used in all our tests in order to use a test npm registry, but consider it unstable.

**Comment by ry** [comment]
Probably we should just use the same env var as NPM: `NPM_CONFIG_REGISTRY`

**Comment by lino-levan** [comment]
https://github.com/denoland/deno/pull/16980/files got merged, should this issue be closed?

**Comment by fufuuuu** [comment]
So how to set NPM_CONFIG_REGISTRY?

**Comment by aapoalas** [comment]
Like you would normally set environment variables. Google may help you with this if you do not know how to do it already.

**Comment by silverwind** [comment]
Ideally deno should support parsing registry and per-registry auth tokens from the [npmrc](https://docs.npmjs.com/cli/v9/configuring-npm/npmrc#files) files.

**Comment by bartlomieju** [comment]
> Ideally deno should support parsing per-scope registry and per-registry auth tokens from the [npmrc](https://docs.npmjs.com/cli/v9/configuring-npm/npmrc#files) files and corsponding env vars, so that users can use multiple registries and tokens.

It's mentioned in a comment above: https://github.com/denoland/deno/issues/16105#issuecomment-1341330077

**Comment by reyawn** [comment]
> We'll look into support for different registries for scopes and credentials in the future.

Is there any update on this @dsherret?   This is a big blocker for many use cases.

**Comment by dsherret** [comment]
The `NPM_CONFIG_REGISTRY` landed in 1.29, but we haven't added support for different registries per scope yet.

**Comment by marziply** [comment]
Where I work, there are some Node microservices that could benefit from migrating to Deno. However, these microservices depend on private Gitlab NPM packages. This feature is blocking the upgrade to Deno because many of these private packages are critical to these services and cannot run without them.

The work so far on this has been excellent and I've not reported any other issues. I will just have to wait until Deno supports private packages before upgrading.

**Comment by marcushultman** [comment]
Implementing this as a [Configuration file](https://deno.com/manual@v1.34.1/getting_started/configuration_file) option (`"npmRegistry"`) in https://github.com/denoland/deno/pull/19317

**Comment by marcushultman** [comment]
@esnho The env was changed to `NPM_CONFIG_REGISTRY`.

**Comment by felipemullen** [comment]
Setting `NPM_CONFIG_REGISTRY` *almost* does the job. In my particular case, there are scoped packages sitting in an artifactory registry that **does not proxy npm** directly. This results in two scenarios:

**`NPM_CONFIG_REGISTRY` is set to my artifactory registry**

- `@myorg/my-package` is resolved correctly ✔️
- global npm packages such as `express` are not ❌

**`NPM_CONFIG_REGISTRY` is not set**

- `@myorg/my-package` is not resolved correctly ❌
- global npm packages such as `express` are fine ✔️

Is there a way I could have two registries set? I tried using spaces ` ` and semicolons `;` in the string, no success. My artifactory uses an auth user with token, which is defined in `.npmrc`

**Comment by bhb603** [comment]
@felipemullen I have the same issue as you...

One workaround I found was to use a local proxy like [Verdaccio](https://verdaccio.org/) to proxy the various registries you need.

**Caveat**: I don't have prior experience with Verdaccio, or setting up my own local npm proxy; I just discovered it when looking for solutions to this multi-registry limitation in Deno 😄 

Anyway, I have Verdaccio configured with the following uplinks and package resolutions:
```
uplinks:
  npmjs:
    url: https://registry.npmjs.org/
  github_private:
    url: https://npm.pkg.github.com/
    auth:
      type: bearer
      token_env: GITHUB_TOKEN
packages:
  '@myorg/*':
    access: $all
    publish: $authenticated
    unpublish: $authenticated
    proxy: github_private

  '**':
    access: $all
    publish: $authenticated
    unpublish: $authenticated
    proxy: npmjs
```
So all `@myorg` scoped packages will pull from Github private registry, and the rest will pull from npm.

Then run `verdaccio` locally, supplying any necessary auth tokens:
```
GITHUB_TOKEN=*** verdaccio
```
Then set the Verdaccio server (localhost:4873 by default) as `NPM_CONFIG_REGISTRY` and use Deno to cache your dependencies:
```
NPM_CONFIG_REGISTRY=http://localhost:4873 deno cache ./my_deno_module.ts
```

**Comment by andykais** [comment]
throwing another monkey wrench into this discussion. I have a use case to retrieve a npm package that is built, and then _only_ available via github. Our company has been able to use github tags and github ssh keys for standard nodejs usage. We use a dependency string that looks like this:
```json
{
  "dependencies": {
    "mypackage": "git+ssh://git@github.com:myorganization/mypackage.git#3.1.1",
  }
}
```

this is part of npm's capabilities, but attempting something similar in deno gives me errors:
```ts
import * as mypackage from 'npm:git+ssh://git@github.com:myorganization/mypackage.git#3.1.1'
```
```
error: Invalid package specifier 'npm:git+ssh://git@github.com:myorganization/mypackage.git#3.1.1'. Did you mean to write 'npm:git+ssh:@github.com:myorganization/mypackage.git#2.1.1//git'?
```

is this part of the npm spec supported by deno?

**Comment by andykais** [comment]
just throwing another monkey wrench into this discussion. I have a use case that involves using npm git dependencies. My company has private npm packages that we simply create git tags for and host on github. These are then defined in package.json like so:
```json
{
  "dependencies": {
    "mypackage": "git+ssh://git@github.com:myorganization/mypackage.git#3.1.1"
  }
}
```

The packages themselves contain native code, and use the npm script `"install": "node-gyp rebuild"` to compile them. This works well for us, the only requirement is that we have a relevant git ssh key added wherever the package needs to be installed. In deno however, this does not appear to work:
```ts
import * as mypackage from 'npm:git+ssh://git@github.com:myorganization/mypackage.git#3.1.1'
```

This results in the following error:
```
error: Invalid package specifier 'npm:git+ssh://git@github.com:myorganization/mypackage.git#3.1.1'. Did you mean to write 'npm:git+ssh:@github.com:myorganization/mypackage.git#3.1.1//git'?```

Is the git ssh reference part of the npm spec supported in deno's npm specifiers?

**Comment by felipemullen** [comment]
Thanks @bhb603, that was a great tip. I was able to get set up using docker-compose to support a self-hosted [windmill.dev](https://www.windmill.dev/docs/advanced/self_host) instance.

For anyone else struggling with the same, here are some notes:

`docker-compose.yaml`
```docker-compose
services:
  verdaccio:
    container_name: verdaccio
    image: verdaccio/verdaccio
    environment:
      - VERDACCIO_PORT=4873
    ports:
      - 4873:4873
    expose:
      - 4873
    volumes:
      - './path/to/local/config:/verdaccio/conf'

windmill_worker:
    ...
    environment:
      - NPM_CONFIG_REGISTRY=http://verdaccio:4873
```

`my/local/verddaccio/config/config.yaml`
```yaml
uplinks:
  npmjs:
    cache: false
    url: https://registry.npmjs.org/
  artifactory:
    cache: false
    url: https://mycorp.artifactory.corp.com/api/npm/my-packages/
    headers:
      Authorization: bearer ABC123ZYX098===AABBCOLLISION

packages:
  '@mycorp/*':
    access: $all
    publish: $authenticated
    unpublish: $authenticated
    proxy: artifactory

  '**':
    access: $all
    publish: $authenticated
    unpublish: $authenticated
    proxy: npmjs
```

Docs references:
- private npm registries for windmill https://www.windmill.dev/docs/advanced/imports#private-npm-registry
- docker networking reference https://docs.docker.com/compose/networking/
- verdaccio docker-compose docs https://verdaccio.org/docs/docker#using-docker-compose

**Comment by dehrhard** [comment]
I second this, I want this so bad. We wouldn't even need scoped support for other registries. Since github is proxying the requests through to npmjs we would only need support for auth with github.

**Comment by nitehogg** [comment]
Is this currently being worked on, or has a decision been made towards the implementation yet?

**Comment by rattrayalex** [comment]
As a package maintainer, I'd like to be able to test packages I'm developing on Deno before publishing. 

The following (or some equivalent) would all be quite useful for being able to offer a great experience to Deno users of my package:
```ts
import mypackage from 'npm:./some-bundle.tar.gz'
import mypackage from 'npm:./some-dir'
import mypackage from 'npm:github.com/my-org/my-repo#v0.2.1-beta'
```

**Comment by bartlomieju** [comment]
Update: we are now working on supporting `.npmrc` file that will allow to use different registries for different scopes.

**Comment by bartlomieju** [comment]
Update: We're gonna land first pass in #23560 really soon.

The first pass will support `.npmrc` located next to `package.json` or `deno.json` - so it will work with Node.js first projects as well as Deno first projects. Please let me know if you rely on `.npmrc` in your user directory or global directory (https://docs.npmjs.com/cli/v10/configuring-npm/npmrc#files).

Only `_authToken` config will be supported in the first pass - please let me know by commenting if you are heavy users of other options so we can prioritize accordingly.

Also the first pass will only support `deno run` / `deno test` subcommands - the support for `deno compile` will come afterwards - please also comment if you need it.

Thanks

**Comment by TaxBusby** [comment]
For Github private registries, we use local-directory `.npmrc`, but notably use env-var interpolation features supported by NPM.

```
@our-github-org:registry=https://npm.pkg.github.com/
//npm.pkg.github.com/:_authToken=${GITHUB_NODE_TOKEN}
```
Developers have this env var set to a valid token.

For Codeartifact private regsitries, we use the [`aws codeartifact login`](https://docs.aws.amazon.com/cli/latest/reference/codeartifact/login.html) tool, which populates `$HOME/.npmrc` with all relevant scope+auth config.

(For adopting Deno, as long as we can globally define the auth token for all repos somewhere, and locally or globally set a scope for the registry, we can figure out a migration path)

**Comment by MirKml** [comment]
Hi, for Azure Artifacts npm - user level .npmrc with the authToken is desired, see https://learn.microsoft.com/en-us/azure/devops/artifacts/get-started-npm?view=azure-devops&tabs=Other

I think this is common scenario for npm scoped registry settings - committed .npmrc with scope and registry URL, then $HOME/.npmrc with authToken for this registry.

**Comment by jtibble** [comment]
@bartlomieju I also wanted to add that this is critical for using Deno within my company, as we have multiple internal proxied NPM repositories (including a copy of NPM, as well as Sonatype Nexus 3.65.x).

Do you have any ETA for when you'd like to try to land this in a release? I'd like to volunteer to test. Local .npmrc compatibility would be incredible with our URLs/auth specified like:

@dcl-crew:registry=https://nexus3.disney.com/repository/npm/
//nexus3.disney.com/repository/npm/:_auth=xxxxxxx

**Comment by bartlomieju** [comment]
@jtibble this fubctionality should now available in canary. Please try it out and let me know.

**Comment by gengurke** [comment]
@bartlomieju I tested this in a node project running node 20 and vite.

**Deno canary version:** `deno 1.43.6+53606de` (updated today)

I placed the `.npmrc` in `~/.npmrc` and the `root directory` of the project. Here is its content:
```
@fortawesome:registry=https://npm.fontawesome.com/
//npm.fontawesome.com/:_authToken=MYTOKEN
```

If i run `deno task dev` I get the following output:

```
error: npm package '@fortawesome/pro-regular-svg-icons' does not exist.
```

Installing and running via npm works fine. I hope this helps.

**Comment by bartlomieju** [comment]
@gengurke please do `DENO_FUTURE=1 deno install` before running `DENO_FUTURE=1 deno task dev`. Let me know if that works for you.

**Comment by jtibble** [comment]
@bartlomieju I also validated with canary build deno 1.43.6+8b5089e (canary, aarch64-apple-darwin) and it worked like a charm! 🙌 

SO excited to have this working locally. What a dream! I have both my registry and auth in my local directory .npmrc like this:

registry=https://nexus3.disney.com/repository/npm/
//nexus3.disney.com/repository/npm/:_auth=b1lDRnZxxxxxxxxxxx

so my only request now is to allow for joining .npmrcs, so I can keep my auth token separate. Not a big deal either way, because we can just .gitignore this, but yes, thanks again! So excited to see this land in the main build!

**Comment by jtibble** [comment]
@bartlomieju Is there any chance the .npmrc functionality currently in the canary builds could be added to the 1.44 release milestone? Even behind a feature flag or toggle would be great, so we can at least start using it on my development team. Thanks!

**Comment by bartlomieju** [comment]
@jtibble, yes this functionality will be in v1.44 and will be announced in a blog post. I will try to get merging of `npmrc` done too, but no promises.

**Comment by jtibble** [comment]
@bartlomieju I also wanted to bring to your attention that in VSCode, with the Deno language server extension, even with the checkbox enabled for Deno: Future, the private-registry-module-resolution is not resolving. This is with canary build 1.43.6+a89253. 

I tried caching through the VSCode popup, no luck. Manually caching the dependency in the terminal works, but still the VSCode extension can't resolve it.


Screenshots:

![Screenshot 2024-05-29 at 11 09 41 AM](https://github.com/denoland/deno/assets/2126464/ef8119c0-89a9-4835-af4c-0b70aba57931)
![Screenshot 2024-05-29 at 11 10 16 AM](https://github.com/denoland/deno/assets/2126464/0b1249a4-77a3-4b0c-bc41-f885eb2c121a)

**Comment by bartlomieju** [comment]
@jtibble thanks for pointing this out, I knew I forgot about something! I opened https://github.com/denoland/deno/issues/24040 and we'll aim to fix it in v1.44.1.

**Comment by sithumonline** [comment]
Worked well with `gitlab` npm registry.

Deno version: 1.43.6+3c3076a
npmrc:
```
@arimac-spades:registry=https://gitlab.com/api/v4/projects/<number>/packages/npm/
//gitlab.com/api/v4/projects/<number>/packages/npm/:_authToken=<token>
```

**Comment by bartlomieju** [comment]
@jtibble looks like it got fixed in time for v1.44: https://github.com/denoland/deno/pull/24042

**Comment by bartlomieju** [comment]
Support for private registries/`.npmrc` has shipped in Deno v1.44.0: https://deno.com/blog/v1.44#support-for-private-npm-registries.

There are a few remaining features that need to be addressed:
- https://github.com/denoland/deno/issues/23953
- https://github.com/denoland/deno/issues/23952
- https://github.com/denoland/deno/issues/23951
- https://github.com/denoland/deno/issues/23950
- https://github.com/denoland/deno/issues/23954

Overall, I think we can now close this issue. Please subscribe to relevant issue or open new ones if you experience any problems.

**Comment by dsherret** [comment]
Hi everyone, we discovered an issue in .npmrc support in Deno 1.44.0 where Deno would send npmrc credentials for the scope to the tarball URL when the registry provided URLs for a tarball on a different domain.

For example, the npm registry provides information on a package at a url like this: https://registry.npmjs.org/code-block-writer -- you can see that under `versions -> <version> -> dist -> tarball` there's a URL. That URL can be anything and Deno was only considering the first URL of https://registry.npmjs.org/code-block-writer to get credentials instead of the tarball URL (so in other words, it would send the credentails for `https://registry.npmjs.org/` to whatever the URL was)

You can read more about this here: https://github.com/npm/cli/wiki/%22No-auth-for-URI,-but-auth-present-for-scoped-registry%22 (how Deno is working is actually how npm used to behave)

I recommend to stop using .npmrc support in Deno 1.44.0 and if your private registry ever serves tarballs at a different domain to rotate your registry credentials. We have a fix for this here: https://github.com/denoland/deno/pull/24111 and are going to do a release in a few hours to fix this issue.

Huge apologies on this one.
